patterns/singleton/singleton.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package singleton
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var _ manager.Runnable = runnable{}
+var _ manager.LeaderElectionRunnable = runnable{}
+
+type runnable struct {
+	objs []client.Object
+	c    client.Client
+}
+
+// NewRunnable returns a manager.Runnable that requires leader election to
+// create all objs using c. This runnable should be added to a manager.Manager
+// with Manager.Add(runnable).
+//
+//	const singletonFooName = "global-foo"
+//
+//	func main() {
+//
+//		mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{})
+//		if err != nil {
+//			os.Exit(1)
+//		}
+//
+//		// Some internally or externally defined API type.
+//		foo := &foosv1alpha1.Foo{}
+//		foo.SetName(singletonFooName)
+//		foo.Spec.Bar = "baz"
+//		// foo will be created after leader election has started.
+//		mgr.Add(singleton.NewRunnable(mgr.GetClient(), foo))
+//
+//	}
+func NewRunnable(c client.Client, objs ...client.Object) manager.Runnable {
+	return runnable{c: c, objs: objs}
+}
+
+func (r runnable) NeedLeaderElection() bool { return true }
+
+// TODO(estroz): parallelize
+func (r runnable) Start(ctx context.Context) error {
+	for _, obj := range r.objs {
+		if err := r.c.Create(ctx, obj); err != nil {
+			return err
+		}
```

Comment:
What happens if these objects are deleted? Since the operator creates them, it seems it would be in-scope for it to re-create them and patch them back to some desired state. Perhaps a `SingletonReconciler` that makes sure the desired objects are always present.

Reply:
@joelanford I'm thinking that [`controllerutil.CreateOrUpdate()`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil?utm_source=godoc#CreateOrUpdate) should be used directly rather than writing a reconciler here, as the reconciler would basically just be calling this function.